ruledata.go

```diff
@@ -814,4 +818,166 @@ func SetRules(s *NetlinkConnection, content []byte) error {
 		}
 	}
 	return nil
-}
\ No newline at end of file
+}
+
+var errPathTooBig = errors.New("The path passed for the watch is too big")
+var errPathStart = errors.New("The path must start with '/'")
+var errBaseTooBig = errors.New("The base name of the path is too big")
+
+func check_path(path_name string) error {
+	if len(path_name) >= PATH_MAX {
+		return errPathTooBig
+	}
+	if path_name[0] != '/' {
+		return errPathStart
+	}
+
+	base := path.Base(path_name)
+
+	if len(base) > syscall.NAME_MAX {
+		return errBaseTooBig
+	}
+
+	if strings.ContainsAny(base, "..") {
+		log.Println("Warning - relative path notation is not supported!")
+	}
+
+	if strings.ContainsAny(base, "*") || strings.ContainsAny(base, "?") {
+		log.Println("Warning - wildcard notation is not supported!")
+	}
+
+	return nil
+}
+
+func AuditSetupWatchName(rule *AuditRuleData, path_name string) error {
+	type_name := uint16(AUDIT_WATCH)
+
+	err := check_path(path_name)
+	if err != nil {
+		return err
+	}
+
+	// Trim trailing '/' should they exist
+	strings.TrimRight(path_name, "/")
+
+	if fileInfo, err := os.Stat(path_name); err != nil {
+		if os.IsNotExist(err) {
+			return errors.New("File does Not exist!")
+		} else {
+			return err
+		}
+		if fileInfo.IsDir() {
+			type_name = uint16(AUDIT_DIR)
+		}
+	}
+
+	err = AuditAddWatchDir(type_name, rule, path_name)
+	if err != nil {
+		return err
+	}
+	return nil
+
+}
+
+func AuditAddWatchDir(type_name uint16, rule *AuditRuleData, path_name string) error {
+
+	// Check if Rule is Empty
+	// if rule && rule.Field_count {
+	// 	return errors.New("Rule is Not Empty!")
+	// }
+
+	if type_name != uint16(AUDIT_DIR) && type_name != uint16(AUDIT_WATCH) {
+		return errors.New("Invalid Type Used!")
+	}
+
+	rule.Flags = uint32(AUDIT_FILTER_EXIT)
+	rule.Action = uint32(AUDIT_ALWAYS)
+	// set mask
+	// TODO : Setup audit_rule_syscallbyname_data(rule, "all")
+	for i:=0 ; i < AUDIT_BITMASK_SIZE-1; i++ {
+		rule.Mask[i] = 0xFFFFFFFF
+	}
+
+
+	rule.Field_count = uint32(2)
+	rule.Fields[0] = uint32(type_name)
+
+	rule.Fieldflags[0] = uint32(AUDIT_EQUAL)
+	valbyte := []byte(path_name)
+	vlen := len(valbyte)
+	// if fieldid == AUDIT_FILTERKEY && vlen > AUDIT_MAX_KEY_LEN {
+	// 	return errMaxLen
+	// } else if vlen > PATH_MAX {
+	// 	return errMaxLen
+	// }
+	rule.Values[0] = (uint32)(vlen)
+	rule.Buflen = (uint32)(vlen)
+	//Now append the key value with the rule buffer space
+	//May need to reallocate memory to rule.Buf i.e. the 0 size byte array, append will take care of that
+	rule.Buf = append(rule.Buf, valbyte[:]...)
+
+	rule.Fields[1] = uint32(AUDIT_PERM)
+	rule.Fieldflags[1] = uint32(AUDIT_EQUAL)
+	rule.Values[1] = uint32(AUDIT_PERM_READ | AUDIT_PERM_WRITE | AUDIT_PERM_EXEC | AUDIT_PERM_ATTR)
+
+	return nil
+}
+
+func AuditSetupPerms(rule *AuditRuleData, perms string) error {
+	if len(perms) > 4 {
```

Comment:
You didn't mention what do you think on the maintaining the `libaudit.c` function signatures.
As I said merging them will cause problems for any client code expecting those functionality only.